### ISSUE TITLE
Rule proposal: `prefer-regexp-unicode-property-escapes`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#988]
Enforce using regex [Unicode property escapes](http://2ality.com/2017/07/regexp-unicode-property-escapes.html) instead of classic escapes whenever possible, as Unicode property escapes improve readability.

More info: https://mathiasbynens.be/notes/es-unicode-property-escapes

`/^\s+$/u` → `/^\p{White_Space}+$/u`

`/^\d+$/u` → `/^\p{Decimal_Number}+$/u`

Originally posted here: https://github.com/eslint/eslint/issues/11382

This rule should probably use https://github.com/DmitrySoshnikov/regexp-tree for the regex parsing.

## Fail

```js
/^\s+$/u;
```

## Pass

```js
/^\p{White_Space}+$/u;
```

**Comment by fregante** [comment]
I'm not convinced that lengthening a regex by 10 characters per instance is a win for readability. I'd enforce this only when there's actually a difference between the regular and unicode escapes.

**Comment by sindresorhus** [comment]
@fisker Any thoughts on this?

**Comment by fisker** [comment]
I think I'll never use `\p{}`, as I understand, no one can remember how many characters that `Decimal_Number` included, and it may change if unicode release new version. Correct me if I'm wrong.

In this example,

```js
const regex = /^\p{Decimal_Number}+$/u;
regex.test('𝟏𝟐𝟑𝟜𝟝𝟞𝟩𝟪𝟫𝟬𝟭𝟮𝟯𝟺𝟻𝟼');
// → true
```

You may think you'll never meet those characters, but in China people used a  lot, but most time people didn't mean to input them, some keyboard are just weird, and no web developer in China treat them as a number.

**Comment by fisker** [comment]
I'm not near my laptop. I don't know if `一` `壹` (Both mean "ONE") are numbers? But we never treat them as number in program .

**Comment by sindresorhus** [comment]
I was under the incorrect assumption that `\d` is the same as `\p{Decimal_Number}` when the `u` flag is used. This rule does indeed not make sense.
